java-frontend/src/main/java/org/sonar/java/cfg/CFGLoop.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.cfg;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.sonar.java.cfg.CFG.Block;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class CFGLoop {
+
+  private final CFG.Block startingBlock;
+  private final Set<CFG.Block> blocks = new HashSet<>();
+  private final Set<CFG.Block> successors = new HashSet<>();
+
+  private CFGLoop(CFG.Block block, Map<Tree, CFGLoop> container) {
+    startingBlock = block;
+  }
+
+  private void initialize(CFG.Block block, Map<Tree, CFGLoop> container) {
+    Block loopFirstBlock = block.trueBlock();
+    if (loopFirstBlock == null) {
+      // Special case where no condition is given in FOR loop: only one successor specified
+      loopFirstBlock = block.successors().iterator().next();
+    }
+    collectBlocks(loopFirstBlock, new HashSet<CFG.Block>(), container);
+    successors.addAll(block.successors());
+    successors.remove(block.falseBlock());
+    collectWaysOut(container);
+  }
+
+  @VisibleForTesting
+  CFG.Block startingBlock() {
+    return startingBlock;
+  }
+
+  @VisibleForTesting
+  Collection<CFG.Block> blocks() {
+    return sorted(blocks);
+  }
+
+  @VisibleForTesting
+  Collection<CFG.Block> successors() {
+    return sorted(successors);
+  }
+
+  public boolean hasNoWayOut() {
+    return successors.isEmpty();
+  }
+
+  private boolean collectBlocks(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
```

Comment:
I'm not sure of the utility of the `@Nonnull` annotation. All our packages are annotated with `@ParametersAreNonnullByDefault` (see `package-info.java`), meaning that the default expectation is that parameters are non-null.